c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,378 @@
+    pub fn should_skip_stmt(&self, stmt: &Statement<'tcx>) -> bool {
+        || -> Option<bool> {
+            let (lhs, rv) = get_assign_sides(stmt)?;
+            let skip = self.to.contains(lhs) || self.from.contains(get_cast_place(rv)?);
+            Some(skip)
+        }()
+        .unwrap_or_default()
+    }
+
+    /// Insert all applicable [`*c_void`] casts
+    /// from the `mir` [`Body`] into this [`CVoidCasts`].
+    ///
+    /// That is, find all calls to:
+    ///
+    /// * `malloc`
+    /// * `calloc`
+    /// * `realloc`
+    /// * `free`
+    ///
+    /// and insert their casts to and from [`*c_void`].
+    ///
+    /// [`*c_void`]: core::ffi::c_void
+    pub fn insert_all_from_mir(&mut self, mir: &Body<'tcx>, tcx: TyCtxt<'tcx>) {
+        for bb_data in mir.basic_blocks().iter() {
+            if let Some(term) = &bb_data.terminator {
+                let term: &Terminator = term;
+                if let TerminatorKind::Call {
+                    ref func,
+                    ref args,
+                    destination,
+                    target,
+                    ..
+                } = term.kind
+                {
+                    let func_ty = func.ty(&mir.local_decls, tcx);
+
+                    // For [`CVoidCastDirection::From`], we only count
+                    // a cast from `*c_void` to an arbitrary type in the subsequent block,
+                    // searching forward.
+                    let find_first_cast_succ_block = |get_cast| {
+                        let successor_block_id = target.unwrap();
+                        mir.basic_blocks()[successor_block_id]
+                            .statements
+                            .iter()
+                            .find_map(get_cast)
+                    };
+
+                    // For [`CVoidCastDirection::From`], we only count
+                    // a cast to `*c_void` from an arbitrary type in the same block,
+                    // searching backwards.
+                    let find_last_cast_curr_block =
+                        |get_cast| bb_data.statements.iter().rev().find_map(get_cast);
+
+                    for direction in CVoidCastDirection::from_callee(ty_callee(tcx, func_ty))
+                        .iter()
+                        .copied()
+                    {
+                        use CVoidCastDirection::*;
+                        let c_void_ptr = match direction {
+                            From => destination,
+                            To => args[0].place().unwrap(),
+                        };
+                        let c_void_ptr = CVoidPtr::checked(c_void_ptr, &mir.local_decls, tcx);
+                        let get_cast = move |stmt: &Statement<'tcx>| {
+                            c_void_ptr.get_cast_from_stmt(direction, stmt)
+                        };
+                        let cast = match direction {
+                            From => find_first_cast_succ_block(get_cast),
```

Comment:
I'm wondering if it's worth keeping the casted and castee in each of `From` and `To` so that things like `args[0].place().unwrap()` only need to be specified in their construction. WDYT?

Reply:
on second thought, it's too minor a point to dwell over